exasol/shared_memory_vault.py

```diff
@@ -0,0 +1,248 @@
+from typing import Optional, Tuple
+from multiprocessing.shared_memory import SharedMemory
+from math import ceil
+from datetime import datetime
+import struct
+
+
+DEFAULT_CRC_DIVISOR = '1001011001100101'
+DEFAULT_MAX_SIZE = 100
+DEFAULT_STORAGE_NAME = 'notebook_connector_vault'
+
+
+def _xor(sequence: str, crc_divisor: str) -> str:
+    return ''.join('0' if a == b else '1' for a, b in zip(sequence, crc_divisor))
+
+
+def compute_crc(sequence: str, crc_divisor: str) -> str:
+    """
+    Computes a Cyclic Redundancy Check (CRC) code for a provided binary sequence using
+    a provided polynomial divisor.
+    Check this wiki for details: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
+    For example if the sequence is '11010011101100000' and the divisor is '011' the
+    output crc code will be '100'. Note that the n+1 bit long divisor, commonly used in
+    literature, 1011 in the above example, is assumed to have the most significant bit
+    (MSB) equal 1. Here the MSB is omitted.
+    """
+    n = len(crc_divisor)
+    reminder = sequence[:n]
+    for i in range(n, len(sequence)):
+        starts_one = reminder[0] == '1'
+        reminder = reminder[1:] + sequence[i]
+        if starts_one:
+            reminder = _xor(reminder, crc_divisor)
+    return reminder
+
+
+def _get_byte_size(n: int) -> int:
+    """Returns the number of bytes required to store an integer value"""
+    return int(ceil(n.bit_length() / 8))
+
+
+def _get_divisor_size(crc_divisor: str) -> int:
+    """Calculates the byte size of a CRC divisor"""
+    return int(ceil(len(crc_divisor) / 8))
+
+
+def _bytes_to_bin(msg_bytes: bytes) -> str:
+    """Converts byte array to a binary string, e.g.[6, 7] => 01100111"""
+    return ''.join(format(c, '08b') for c in msg_bytes)
+
+
+def encode(content: str, crc_divisor: str,
+           creation_time: datetime) -> Tuple[int, bytearray]:
+    """
+    Creates a bytearray with encoded content and its creation datetime.
+    Currently, the content is not being encrypted. It gets appended by the
+    Cyclic Redundancy Check (CRC) code. The CRC is computed over both the timestamp
+    and the content.
+
+    The need for a CRC is debatable. Its use is motivated by the problem of
```

Comment:
Yes, I know, a file lock is usually used as a synchronization mechanism for shared memory. I don't think it's worth doing in this case unless this code finds another usage. Why would we use it in the setup? Currently, the concept is quite simple: 1. The shared memory may or may not contain the data we are after. 2. The data may be corrupted or outdated. The user has to handle both cases, which works well for passwords. We try to distinguish between the first and second cases by using the CRC. But since we cannot do this with 100% accuracy I wonder if it's worth doing it at all.

Reply:
In the worst case, the stuff is currently written, and the user quickly changes to another notebook. You can get simply issues, if you don't do it properly.